equinox_models/RNN.py

```diff
@@ -0,0 +1,105 @@
+import abc
+
+import equinox as eqx
+import jax
+import jax.numpy as jnp
+
+
+class _AbstractRNNCell(eqx.Module):
+    """Abstract RNN Cell class."""
+
+    cell: eqx.Module
+    hidden_size: int
+
+    @abc.abstractmethod
+    def __init__(self, data_dim, hidden_dim, *, key):
+        """Initialize RNN cell."""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def __call__(self, state, input):
+        """Call method for RNN cell."""
+        raise NotImplementedError
+
+
+class LinearCell(_AbstractRNNCell):
+    cell: eqx.nn.Linear
+    hidden_size: int
+
+    def __init__(self, data_dim, hidden_dim, *, key):
+        self.cell = eqx.nn.Linear(data_dim + hidden_dim, hidden_dim, key=key)
+        self.hidden_size = hidden_dim
+
+    def __call__(self, state, input):
+        return self.cell(jnp.concatenate([state, input]))
+
+
+class GRUCell(_AbstractRNNCell):
+    cell: eqx.nn.GRUCell
+    hidden_size: int
+
+    def __init__(self, data_dim, hidden_dim, *, key):
+        self.cell = eqx.nn.GRUCell(data_dim, hidden_dim, key=key)
+        self.hidden_size = hidden_dim
+
+    def __call__(self, state, input):
+        return self.cell(input, state)
+
+
+class LSTMCell(_AbstractRNNCell):
+    cell: eqx.nn.LSTMCell
+    hidden_size: int
+
+    def __init__(self, data_dim, hidden_dim, *, key):
+        self.cell = eqx.nn.LSTMCell(data_dim, hidden_dim, key=key)
+        self.hidden_size = hidden_dim
+
+    def __call__(self, state, input):
+        return self.cell(input, state)
+
+
+class MLPCell(_AbstractRNNCell):
+    cell: eqx.nn.MLP
+    hidden_size: int
+
+    def __init__(self, data_dim, hidden_dim, depth, width, *, key):
+        self.cell = eqx.nn.MLP(data_dim + hidden_dim, hidden_dim, width, depth, key=key)
+        self.hidden_size = hidden_dim
+
+    def __call__(self, state, input):
+        return self.cell(jnp.concatenate([state, input]))
+
+
+class RNN(eqx.Module):
+    cell: _AbstractRNNCell
+    output_layer: eqx.nn.Linear
+    hidden_dim: int
+    classification: bool
+    stateful: bool = False
+    nondeterministic: bool = False
```

Comment:
Are `stateful` and `nondeterministic` here just for API compatibility? They don't seem to be used anywhere. (If so, maybe leave a comment?)

Reply:
These are left over from me building a training loop which can easily swap an RNN out for something that is stateful or nondeterministic, e.g. S5. However, I feel they are probably an unnecessary addition here so happy to remove them.